### ISSUE TITLE
S3 Multi-Region Access Point(MRAP) is not available unless with additional dependency

### ISSUE THREAD

**Issue by AllanZhengYP** [aws/aws-sdk-js-v3#2822]
### Description
Package `@aws-sdk/client-s3` launched the support for S3 Multi-Region Access Point(MRAP) feature since version `3.31.0`. This feature relies on [`aws-crt`](https://github.com/awslabs/aws-crt-nodejs) package's essential signer class. This dependency requires  [native addons](https://nodejs.org/api/addons.html) that takes upto 10.5 MB on disk. Because the dependency is only required by MRAP feature. We decided omit the dependency from the S3 client's dependencies. 

To use this feature, you need to install this dependency explicitly: 

```console
npm install aws-crt
# yarn add aws-crt
```

Because the aws-crt's implementation contains platform-specific binary code, so when you need MRAP feature, the following use cases are not supported:
* browsers
* React Native
* bundle as Node.js application

The suggested use case for supporting MRAP in Node.js app is to deploy your application with the complete `node_modules` folder.

### Describe alternatives you've considered
We are working on solutions to unblock the MRAP for all the supported platforms.

**Comment by seriousme** [comment]
FYI: One might consider porting the signer class of the common runtime to [webassembly](https://webassembly.org/). Web assembly is supported in all major browsers and in NodeJs 8+

Kind regards,
Hans

**Comment by christophgysin** [comment]
Adding aws-crt blocks me from updating the [deno port of aws-sdk-js-v3](https://github.com/christophgysin/aws-sdk-js-v3) past v3.33.0.

:+1: for WASM, which would also allow using this in deno.

**Comment by rossng** [comment]
Just want to weigh in with the impact this issue currently has on our team. Hopefully it helps to motivate a fix. Our developers use a mix of macOS and Windows machines.

We depend on the `@aws-sdk/client-s3` package. This causes the package manager to complain about the missing `@aws-sdk/signature-v4-crt` peer dependency.

The Windows users go ahead and add the dep to `package.json`, and everything appears to be resolved. Then a macOS user pulls the latest changes and suddenly package installation is completely broken.

This isn't a great experience. Given that `@aws-sdk/signature-v4-crt` has ~63k downloads per week, I can only imagine how many developers have run headlong into this so far.

**Comment by jekh** [comment]
I just ran into this issue, seeing the same peer dependency warnings as @rossng. Using the `peerDependenciesMeta` field of package.json might be a good solution: https://docs.npmjs.com/cli/v8/configuring-npm/package-json#peerdependenciesmeta

Adding this to package.json in [middleware-sdk-s3](https://github.com/aws/aws-sdk-js-v3/blob/main/packages/middleware-sdk-s3/package.json) should fix this:
```json
"peerDependenciesMeta": {
   "@aws-sdk/signature-v4-crt": {
     "optional": true
   }
 }
```

**Comment by AllanZhengYP** [comment]
@jekh Thanks a lot for pointing this out! It's a miss on our side. I will address them a separate PR.

**Comment by CraigStuntzWillowTree** [comment]
I just tried this with `3.58.0` and I'm still seeing the error on macOS:
```
WARNING in ../node_modules/@aws-sdk/util-user-agent-node/dist-es/is-crt-available.js 3:96-114
Module not found: Error: Can't resolve 'aws-crt' in '/Users/craig/Documents/PMI/node_modules/@aws-sdk/util-user-agent-node/dist-es'
 @ ../node_modules/@aws-sdk/util-user-agent-node/dist-es/index.js 5:0-52 16:23-37
 @ ../node_modules/@aws-sdk/client-ssm/dist-es/runtimeConfig.js 12:0-65 22:898-914
 @ ../node_modules/@aws-sdk/client-ssm/dist-es/SSMClient.js 10:0-73 15:24-42
 @ ../node_modules/@aws-sdk/client-ssm/dist-es/index.js 2:0-28 2:0-28
...
```
...where I have the library referenced like so:
```json
    "@aws-sdk/client-ssm": "^3.58.0"
```
and it's warning on this line:
```typescript
import { SSM } from "@aws-sdk/client-ssm";
```

**Comment by Hasan-Jawaheri** [comment]
This issue is affecting our team and our entire organization will run into the same issue soon as we are moving towards a DNS regional failover which requires SigV4A signing in the browser to make API calls to unkown AWS region (using `@aws-sdk/signature-v4-crt`).

Are there alternatives for the time being? Can we do SigV4A signing with existing parts of the SDK that work on browser?

**Comment by samfroggy** [comment]
This is insane.
No update from Amazon? This has put an entire build on hold.

Minimal documentation for MRAP as it is.

**Comment by ashley-jisc** [comment]
Still nothing for this?

**Comment by JClackett** [comment]
any updates on this?

**Comment by anta-semenov** [comment]
any updates on this?

**Comment by anta-semenov** [comment]
@AllanZhengYP I'm curious does c++ code has any platform specific dependencies?
It seems like the latest version of react native has first class support c++ code with turbomodules. Does the team consider the option to create a separate rn library for that

**Comment by MRdNk** [comment]
I'm getting the following error in CloudWatch, when trying to do a putObjectCommand to an MRAP...

```json
ERROR	Invoke Error 	
{
    "errorType": "TypeError",
    "errorMessage": "Cannot read properties of undefined (reading 'enable_logging')",
    "$metadata": {
        "attempts": 1,
        "totalRetryDelay": 0
    },
    "stack": [
        "TypeError: Cannot read properties of undefined (reading 'enable_logging')",
        "    at new CrtSignerV4 (/var/task/[bundleName].js:41108:22)",
        "    at SignatureV4MultiRegion.getSigv4aSigner (/var/task/[bundleName].js:41056:33)",
        "    at SignatureV4MultiRegion.sign (/var/task/announcementHandlerBundle.js:41029:25)",
        "    at /var/task/[bundleName].js:9155:31",
        "    at process.processTicksAndRejections (node:internal/process/task_queues:95:5)",
        "    at async /var/task/[bundleName].js:6896:46",
        "    at async /var/task/[bundleName].js:54832:20",
        "    at async /var/task/[bundleName].js:5636:26"
    ]
}```

Using the latest S3 and cert lib., in my package.json ...

```json
"@aws-sdk/client-s3": "^3.391.0",
"@aws-sdk/signature-v4-crt": "^3.391.0",
```

S3Client setup as follows (in side a Ts Lambda function, which works if I just use a standard s3 bucket):

```ts
const s3Client = new S3Client({ useGlobalEndpoint: true });
```

I've tried with and without useGlobalEndpoint; as well as with and without setting the region: 'eu-west-2'; the region the lambda is running in - all the same issue.

Any help appreciated.

**Comment by MRdNk** [comment]
Just read the doesn’t work with bundled nodejs applications; with advice to copy in the node_modules.
That’s not going to be easy

**Comment by MicrosoftJones** [comment]
Patch

**Comment by kuhe** [comment]
The AWS SDK for JavaScript team is planning to change the usage of the optional CRT dependency in a future update.

The change is described in this announcement: https://github.com/aws/aws-sdk-js-v3/issues/5229.

The change is aimed at reducing errors encountered by bundlers and other static analysis tools. The `@aws-sdk/signature-v4-crt` package remains optional and is only required when using S3 MRAP or EventBridge.

**Comment by Legacy131313** [comment]
> ### Description
> Package `@aws-sdk/client-s3` launched the support for S3 Multi-Region Access Point (MRAP) feature in `v3.31.0`. This feature relies on signer class from [`aws-crt`](https://github.com/awslabs/aws-crt-nodejs) package. This dependency requires [native addons](https://nodejs.org/api/addons.html) that takes up to 10.5 MB on disk. Because the dependency is only required by MRAP feature, we are omitting the dependency in `@aws-sdk/client-s3`.
> 
> To use this feature, you need to install this dependency explicitly in your application:
> 
> ```
> $ npm install @aws-sdk/signature-v4-crt
> $ yarn add @aws-sdk/signature-v4-crt
> ```
> 
> Because aws-crt contains platform-specific binary code, the following use cases are not supported in MRAP:
> 
> * browsers
> * React Native
> * bundle as Node.js application
> 
> The suggested use case for supporting MRAP in Node.js app is to deploy your application with the complete `node_modules` folder.
> 
> ### Describe alternatives you've considered
> We are working on solutions to unblock the MRAP for all the supported platforms.
> 
> ### Related issues:
> * [@aws-sdk/client-s3 makes node_modules too large #2747](https://github.com/aws/aws-sdk-js-v3/issues/2747)
> * [@aws-sdk/client-s3 (@aws-sdk/signature-v4-crt) 3.29.0 introduces aws-crt and deprecated dependencies #2750](https://github.com/aws/aws-sdk-js-v3/issues/2750)
> * [AWS CRT binary not present in any of the following locations #2753](https://github.com/aws/aws-sdk-js-v3/issues/2753)
> * [Missing dependency @aws-sdk/signature-v4-crt after updating to client-s3 3.33.0 #2806](https://github.com/aws/aws-sdk-js-v3/issues/2806)
> * [Missing Peer Dependency | @aws-sdk/client-s3 missing @aws-sdk/signature-v4-crt #2960](https://github.com/aws/aws-sdk-js-v3/issues/2960)

**Comment by oskbor** [comment]
So I have spent the last weeks setting up buckets, replication rules, an Multi Region Access Point and I have gotten STS to sign an access token that I have now passed to my client. And here I am, just now finding out that _oh btw, multi region access points cannot be queried from the browser_. Que? I'm fuming right now. 

Isn't that like a very common use case to query s3 from the browser?
At least put this information in some MRAP docs 😢 Very sad to have spent so much time on this.

Please tell me that there is a way to do `GetObject` from a browser via an mrap.

**Comment by siddsriv** [comment]
Functionalities such as S3 MRAP need to use SigV4A as the signing algorithm. The SDK either needs to use the [`aws-crt` implementation](https://github.com/aws/aws-sdk-js-v3?tab=readme-ov-file#functionality-requiring-aws-common-runtime-crt) or the [JS implementation of SigV4A](https://github.com/aws/aws-sdk-js-v3?tab=readme-ov-file#using-javascript-non-crt-implementation-of-sigv4a). Instructions to use them and related notes are present in our [README](https://github.com/aws/aws-sdk-js-v3?tab=readme-ov-file#known-issues).

No further reduction of dependencies for this functionality is planned at this time. Closing this issue, let us know if something comes up!
